proxy/envoy/discovery.go

```diff
@@ -125,101 +184,208 @@ func (ds *DiscoveryService) Run() {
-	// There is a lot of potential to cache and reuse cluster definitions across proxies and also
-	// skip computing the actual HTTP routes
-	instances := ds.services.HostInstances(map[string]bool{ip: true})
-	services := ds.services.Services()
-	httpRouteConfigs := buildOutboundHTTPRoutes(instances, services, &ProxyContext{
-		Discovery:  ds.services,
-		Config:     ds.config,
-		MeshConfig: ds.mesh,
-		IPAddress:  ip,
-	})
+func (ds *DiscoveryService) cachedDiscoveryResponse(key string) ([]byte, bool) {
+	if ds.disableCache {
+		return nil, false
+	}
 
-	// de-duplicate and canonicalize clusters
-	clusters := httpRouteConfigs.clusters().normalize()
+	ds.cacheMu.RLock()
+	defer ds.cacheMu.RUnlock()
 
-	// apply custom policies for HTTP clusters
-	for _, cluster := range clusters {
-		insertDestinationPolicy(ds.config, cluster)
+	// Miss - entry.miss is updated in updateCachedDiscoveryResponse
+	entry, ok := ds.cache[key]
+	if !ok || entry.data == nil {
+		return nil, false
 	}
 
-	if err := response.WriteEntity(ClusterManager{Clusters: clusters}); err != nil {
-		glog.Warning(err)
+	// Hit
+	atomic.AddUint64(&entry.hit, 1)
+	return entry.data, true
+}
+
+func (ds *DiscoveryService) updateCachedDiscoveryResponse(key string, data []byte) {
+	if ds.disableCache {
+		return
+	}
+
+	ds.cacheMu.Lock()
+	defer ds.cacheMu.Unlock()
+
+	entry, ok := ds.cache[key]
+	if !ok {
+		entry = &discoveryCacheEntry{}
+		ds.cache[key] = entry
+	} else if entry.data != nil {
+		glog.Warningf("Overriding cached data for entry %v", key)
+	}
+	entry.data = data
+	atomic.AddUint64(&entry.miss, 1)
+}
+
+// ListEndpoints responds to SDS requests
+func (ds *DiscoveryService) ListEndpoints(request *restful.Request, response *restful.Response) {
+	key := request.Request.URL.String()
+	out, cached := ds.cachedDiscoveryResponse(key)
+	if !cached {
+		hostname, ports, tags := model.ParseServiceKey(request.PathParameter(ServiceKey))
+		// envoy expects an empty array if no hosts are available
+		hostArray := make([]*host, 0)
+		for _, ep := range ds.services.Instances(hostname, ports.GetNames(), tags) {
+			hostArray = append(hostArray, &host{
+				Address: ep.Endpoint.Address,
+				Port:    ep.Endpoint.Port,
+			})
+		}
+		var err error
+		if out, err = json.MarshalIndent(hosts{Hosts: hostArray}, " ", " "); err != nil {
+			errorResponse(response, http.StatusInternalServerError, err.Error())
+			return
+		}
+		ds.updateCachedDiscoveryResponse(key, out)
+	}
+	writeResponse(response, out)
+}
+
+// ListClusters responds to CDS requests for all outbound clusters
+func (ds *DiscoveryService) ListClusters(request *restful.Request, response *restful.Response) {
+	key := request.Request.URL.String()
+	out, cached := ds.cachedDiscoveryResponse(key)
+	if !cached {
+		if sc := request.PathParameter(ServiceCluster); sc != ds.mesh.IstioServiceCluster {
+			errorResponse(response, http.StatusNotFound,
+				fmt.Sprintf("Unexpected %s %q", ServiceCluster, sc))
+			return
+		}
```

Comment:
I considered this but decided against it because that optimizes for the uncommon / slow-path case where `ServiceCluster != istio-proxy`. I don't think there is currently any case where this is true?